src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,233 @@
+   ;; change in the future, thus -motions
+   :KeyJ        [:relative-position (:down  -motions)]
+   :KeyK        [:relative-position (:up    -motions)]
+   :KeyH        [:relative-position (:left  -motions)]
+   :KeyL        [:relative-position (:right -motions)]
+
+   :KeyEnter    [:relative-position (:down  -motions)]
+   :ArrowDown   [:relative-position (:down  -motions)]
+   :ArrowUp     [:relative-position (:up    -motions)]
+   :ArrowLeft   [:relative-position (:left  -motions)]
+   :ArrowRight  [:relative-position (:right -motions)]
+
+   :KeyW        [:relative-position [measure-size     0 0]]
+   :KeyB        [:relative-position [(- measure-size) 0 0]]
+
+   :ShiftBracketRight [:frame  1]
+   :ShiftBracketLeft  [:frame -1]
+
+   ;; absolute movement
+   :KeyG        [:absolute-position [:first :first :first]]
+   :ShiftKeyG   [:absolute-position [:last  :last  :last]]
+   :Digit0      [:absolute-position [:first :first :first]]
+   :ShiftDigit4 [:absolute-position [:last  :last  :last]]
+
+   ;; mode change
+   :KeyI [:mode :edit]
+   :KeyV [:mode :visual]
+
+   ;; other
+   :KeyX        [:macro {:type :simple :value [[:attr nil] (:down-line -motions)]}]
+   ;; so these could be special or just, like... :tempo, and :octave
+   ;; I'm leaving them out for now
+   ;; :Period [:special [:tempo 1]]
+   ;; :ShiftPeriod [:special [:tempo -1]]
+   :Space [:special :play-pause]})
+
+(def edit-keymap
+  ;; hm, this isn't going to work quite like you want it to.
+  ;; this is fine for inputting notes...
+  {:KeyA [:attr :C]    :KeyW [:attr :C#]
+   :KeyS [:attr :D]    :KeyE [:attr :D#]
+   :KeyD [:attr :E]    :KeyR [:attr :F]
+   :KeyF [:attr :F#]   :KeyT [:attr :G]
+   :KeyG [:attr :G#]   :KeyY [:attr :A]
+   :KeyH [:attr :A#]   :KeyU [:attr :B]
+
+   ;; ... but when you get to the other attributes, what happens? They have
+   ;; values in [0x00, 0xff].
+   ;; You gotta either have another sort of translation layer where some of your
+   ;; note names become hex digits, or... something
+   ;; In fact, now that I think about it, this is why volume in the pre-alpha version
+   ;; only goes up to 9.
+   :Digit0 [:attr :Digit0]
+   :Digit1 [:attr :Digit1]
+   :Digit2 [:attr :Digit2]
+   :Digit3 [:attr :Digit3]
+   :Digit4 [:attr :Digit4]
+   :Digit5 [:attr :Digit5]
+   :Digit6 [:attr :Digit6]
+   :Digit7 [:attr :Digit7]
+   :Digit8 [:attr :Digit8]
+   :Digit9 [:attr :Digit9]
+
+   :KeyX      :off
+   :ShiftKeyX :stop
+
+   :Backspace [:macro {:type :simple :value [[:attr nil] (:up-line -motions)]}]
+   :Space     [:relative-position (:down-line -motions)]})
+
+(def mode-keymaps
+  {:normal normal-keymap
+   :edit   edit-keymap})
+
+(def -movement-keys
+  #{:Space :ArrowDown :ArrowUp :ArrowLeft :ArrowRight :Tab :Backspace})
+
+(defn prevent-movement! [ev]
+  "Prevent movement keys (arrows, space, tab) from moving focus or scrolling."
+  (when (-movement-keys (keyword (.-code ev)))
+    (.preventDefault ev))
+  ev)
+
+(defn translate-keycode [ev]
```

Comment:
Rename this to `-translate-keycode`.

The `-` is supposed to denote a namespace-internal value, but really it's just me indicating something is a hack.